modules/ldapbackend/powerldap.cc

```diff
@@ -324,6 +442,51 @@ PowerLDAP::SearchResult::Ptr PowerLDAP::search( const string& base, int scope, c
 
   return SearchResult::Ptr( new SearchResult( msgid, d_ld ) );
 }
+ 
+ 
+PowerLDAP::SearchResult::Ptr PowerLDAP::sorted_search( const string& base, int scope, const string& filter, const string& sort, const char** attr, unsigned int limit )
+{
+  int msgid, rc, sort_rc;
+
+  if ( !d_sort_supported )
+    return SearchResult::Ptr();
+
+  LDAPControl* sortcontrol;
+  LDAPSortKey** sortkey;
+  ldap_create_sort_keylist( &sortkey, const_cast<char*>( sort.c_str() ) );
+  rc = ldap_create_sort_control( d_ld, sortkey, 1, &sortcontrol );
+  if ( rc != LDAP_SUCCESS )
+    throw LDAPException( "Failed to create sort control: " + getError( rc ) );
+  ldap_free_sort_keylist( sortkey );
+
+  LDAPControl* vlvcontrol = NULL;
+  if ( d_vlv_supported && limit > 0 ) {
+    LDAPVLVInfo virtuallist;
+    virtuallist.ldvlv_version = 1;
+    virtuallist.ldvlv_before_count = 0;
+    virtuallist.ldvlv_after_count = limit;
+    virtuallist.ldvlv_offset = 0;
+    virtuallist.ldvlv_count = 1;
+    ldap_create_vlv_control( d_ld, &virtuallist, &vlvcontrol );
+  }
+
+  LDAPControl* servercontrols[3];
+  servercontrols[0] = sortcontrol;
+  servercontrols[1] = vlvcontrol;
+  servercontrols[2] = NULL;
+
+  rc = ldap_search_ext( d_ld, base.c_str(), scope, filter.c_str(), const_cast<char**> (attr), 0, servercontrols, NULL, NULL, LDAP_NO_LIMIT, &msgid );
+  if ( rc == LDAP_SERVER_DOWN || rc == LDAP_CONNECT_ERROR )
+    throw LDAPNoConnection();
```

Comment:
Don't we leak `sortcontrol` and `vlvcontrol` when we throw here and right below?